### PR TITLE
Add optional init_block kwarg to build_tarball

### PR DESCRIPTION
**Pull request by fingolfin**
This optional argument can be set to a string, which is then inserted at
the end of the generated __init__ methods. This allows inserting code
which initializes the JLL or in other ways makes it easier to use; e.g
for ncurses, it could set the TERMINFO env variable; for GMP, it could call
the `mp_set_memory_functions` function to ensure this copy of GMP uses
the same alloc/free functions as the GMP bundled with Julia.

This is motivated by issue #687 and would e.g. allow us to get rid of https://github.com/oscar-system/LoadFlint.jl (by changing https://github.com/JuliaBinaryWrappers/FLINT_jll.jl to contain the appropriate initialization calls).

Full disclaimer: the code in here is 100% untested! So it may even contain basic syntax errors. The reasons for this is that while I am willing to spend time making it work etc., I also don't want to waste time figuring everything out from scratch if I then might be told later on that the whole approach is meritless and won't ever have a chance of being merged :-)

Hence, I thought I'd first solicit some basic feedback. E.g.
- is the idea of inserting code into `__init__` in principle OK? Perhaps the code should be inserted elsewhere (many variations come to mind; e.g. it could be a single function defined in the main file, which then is invoked by `__init__`; etc.)
- is the same `init_block` OK? Other suggestions?
- it OK that I am simply inserting a string? The rationale is that this is dead easy to do. In issue #687, @simonbyrne suggested that one could also pass a `quote` block -- I can also look into this, but it seems to be more complex with no obvious (to me) gain. But I'll do whatever the consensus is.
- I wanted to add tests, but I found no existing tests for `build_jll_package` or `build_tarballs` so I don't know how I should go about that...

### CODE REVIEW

**Comment by fingolfin** [comment]
@giordano If you have any thoughts on this, I'd appreciate your input, if your time permits.

**Comment by staticfloat** [comment]
I was hoping to avoid this. 😞

I've been hoping against hope that binary packages can be mostly loaded with a very cookie-cutter approach, so that we can identify how most binary packages can be loaded, then do things like eliminate JLL packages completely and have that binary loading logic stored within Julia base itself.  The less complexity and the more regularity we have in code loading, the faster and more reliable it will be.

Allowing arbitrary code to be packed into the `__init__()` method of your JLL is only going to increase the surface area of things that can go wrong and cause you problems.  However, you raise great points that not allowing this makes life very difficult in certain strange situations.  Therefore, I'm open to adding this as an escape hatch, but I think we should be very careful in using it, and we should do our best to add more structured methods that make dealing with classes of these problems easier.  Examples:

* For the environment variable setting issues in https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/687, we could have users provide products that should get mapped to environment variables that get set automatically upon loading a JLL.  Something that looks like:

```julia
build_tarballs(...; jll_environment_vars = Dict(
    # Mapping to a string sets a constant value across all platforms
    "CONSTANT_VAR" => "This environment variable is always set to the same thing",
    # Mapping to a product sets the result of `locate()`'ing the product
    "LIBFOO_PATH" => LibraryProduct("libfoo", :libfoo),
))
```

This is probably a good idea, and I think could be implemented now.

* For the need to call initialization functions.... I'm not sure what a more structured API would look like.  Perhaps this can be integrated into our ideas for a [simple JLL test suite](https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/647) as well; we're already thinking about how to embed pieces of code into the `build_tarballs.jl`.  I'm not sure.

> I wanted to add tests, but I found no existing tests for `build_jll_package` or `build_tarballs` so I don't know how I should go about that...

.....yeah, that's a long-standing dev wishlist item.  ;)

**Comment by fingolfin** [comment]
> this makes life very difficult in certain strange situations

I don't think those are strange situations; depending on what you do, they are common!

E.g. Gmp_jll does not init GMP to use Julia's allocation functions. Hence one must be very careful to either initialize it very, very early (but that may be impossible if multiple packages using it are involved, and not all are under your control), or else generally ensure that the shared library in there is not actually used (and that Julia's bundled GMP is used instead), otherwise one can run into nasty situations if one ever wants to use a GMP integer allocated in one copy of GMP inside another copy -- segfaults or worse are otherwise likely.

Anyway, I totally understand your wish, but I am concerned that trying to find the perfect solution here will delay implementing *any* solution indefinitely. I understand this feature can be easily abused, and use should be minimized, but *some* escape hatch should be there.


So, how can we move forward? Sorry, the env var setting (which indeed sounds useful) simply isn't enough for my usecases.


Note that I did think about alternatives. e.g. for GMP and Flint, one could also apply a patch to the source code which changes the default memory allocation functions to use those of Julia. The problem with that is that it requires linking to Julia; of course now we have Julia_JLL; but during runtime, we *really* want to link against the active Julia, not another copy. That may be possible (I did not try this), but my point here is that this is rather non-trivial.

For the env variables, in some cases it might also be possible (and even better) to patch source code to insert different default paths that work under relocation; I did such things for a different kind of "reloctable binary distribution" I created years ago (nothing fancy, see https://github.com/fingolfin/gap-osx-bundle). But that particular one was much easier than what you are doing here, in that it only targeted macOS, and everything was in an `.app` bundle, so I could leverage OS support to make this easier; getting this kind of patch to work across Linux, macOS, Windows, FreeBSD is highly non-trivial, and one must be wary of accidentally breaking the patched code; still, perhaps it could be used in some cases, esp. if BinaryBuilder provides some help. I only have some fuzzy ideas on what that "help" might look like; really, the best way is as you describe it: attempt to do it, then watch out for patterns that you can factor out.

I mean, you already do some of that by tweaking the shared libraries to be relocatable; but e.g. for shell scripts, this might also be useful.

BTW, some JLLs contain hardcoded paths derived from the build system. This is usually a source of non-relocatability. In https://github.com/fingolfin/gap-osx-bundle I thus added extra checks which grep through the produced binaries for the builddir (which had a very unique name to avoid accidental hits), and produced warnings/errors in that case. I then tried hard to get rid all of these (to be "warning free", and also to not leak info about the build system, which is an important step towards [reproducible builds](https://reproducible-builds.org), something I hope JLLs will slowly converge towards).

**Comment by fingolfin** [comment]
@staticfloat it would be really good to make progress on this. Right now, though I don't know what you think needs to be done before this can be merged? I understand that you'd prefer a "better" solution to the initialization problem, but I don't know what that's supposed to look like, so I can't work on it either. 

I agree that a way to just set env vars would be desirable, and it also seems quite doable (see also https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/687#issuecomment-635867414). But that doesn't solve the issue of having to call some initialization APIs...

Perhaps @giordano has some insights?

**Comment by staticfloat** [comment]
@fingolfin Alright, your persistence has paid off. :)

Rebase this, and I'll merge the `init_block` as a string being inserted into the end of `__init__()`.  I would like you to document it as an experimental feature that can disappear in the future.  I would like to work toward having better APIs that don't require arbitrary code execution (or if they do, to have it in a slightly more user-friendly manner, e.g. a bundled `.jl` file that gets included from within `__init__()` perhaps?) but that shouldn't block you from getting real work done.

**Comment by fingolfin** [comment]
I've rebased this now. Marking this as experimental seems sensible; I'll be happy to adjust if / once "better" approaches are found. 

In the meantime, being able to watch how some packages use `init_block` might be helpful in figuring out alternative approaches.

**Comment by fingolfin** [comment]
Ahh, before this is merged, I should point out that I still did not actually test this thoroughly, partially because I am not sure *how* to test it.

File: src/AutoBuild.jl [hunk 1]
```diff
@@ -920,6 +930,7 @@ function rebuild_jll_package(obj::Dict;
         upload_prefix;
         verbose=verbose,
         lazy_artifacts = lazy_artifacts,
+        init_block = get(obj, "init_block", ""),
```
Comment (fingolfin) [inline src/AutoBuild.jl:933]: Is this the correct way to do it?

File: src/AutoBuild.jl [hunk 2]
```diff
@@ -982,15 +993,15 @@ function rebuild_jll_package(name::String, build_version::VersionNumber, sources
...
+        # If `from_scratch` is set (the default) we clear out any old crusty code
         # before generating our new, pristine, JLL package within it.  :)
         if from_scratch
             rm(joinpath(code_dir, "src"); recursive=true, force=true)
             rm(joinpath(code_dir, "Artifacts.toml"); force=true)
         end
 
         # Finally, generate the full JLL package
-        build_jll_package(name, build_version, sources, code_dir, build_output_meta, dependencies, upload_prefix; verbose=verbose, lazy_artifacts=lazy_artifacts)
+        build_jll_package(name, build_version, sources, code_dir, build_output_meta, dependencies, upload_prefix; verbose=verbose, lazy_artifacts=lazy_artifacts, init_block=init_block)
```
Comment (fingolfin) [inline src/AutoBuild.jl:1004]: Perhaps this line should be wrapped? Are you using `JuliaFormatter` or something like that?

**Comment by fingolfin** [comment]
@staticfloat  build passes, but I marked one spot where I am not 100% sure. Also, should I added comments indicating "this is experimental"?

File: src/AutoBuild.jl (hunk 2 above)
Comment (staticfloat) [reply src/AutoBuild.jl:1004]: No, but I usually try to wrap things at about 90 characters.  So yes please, feel free to wrap it.

File: src/AutoBuild.jl (hunk 1 above)
Comment (staticfloat) [reply src/AutoBuild.jl:933]: We can assume that the only `obj` objects coming in are ones generated by this same BB version; so we will always have an `init_block` key in `obj`.

File: src/AutoBuild.jl [hunk 3]
```diff
@@ -76,6 +76,12 @@ download, build and package the tarballs, generating a `build.jl` file when
 appropriate.  Note that `ARGS` should be the top-level Julia `ARGS` command-
 line arguments object.  This function does some rudimentary parsing of the
 `ARGS`, call it with `--help` in the `ARGS` to see what it can do.
+
+The `kwargs` are passed on to [`autobuild`](@ref), see there for a list of
+supported ones. In addition, the keyword argument `init_block` may be set to
+a string containing Julia code; if present, this code will be inserted into
+the initialization path of the generated JLL package. This can for example be
+used to invoke an initialization API of a shared library.
```
Comment (staticfloat) [inline src/AutoBuild.jl:84]: This is where you should mention that `init_block` is experimental and may be removed in a future release.

File: src/AutoBuild.jl (hunk 2 above)
Comment (fingolfin) [reply src/AutoBuild.jl]: Done. (Lots of lines > 90 in the file, but ... *shrug* ;-)

File: src/AutoBuild.jl (hunk 3 above)
Comment (fingolfin) [reply src/AutoBuild.jl:84]: Done via `!!! note`, let me know if you prefer something else.

**Comment by staticfloat** [comment]
I manually tested this by making this change to the [`HelloWorldC` build script](https://github.com/JuliaPackaging/Yggdrasil/blob/master/H/HelloWorldC/build_tarballs.jl):

```diff
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+               init_block="""
+               ENV["JULIA_HELLO_WORLD"]="C"
+               """,
+               )
```

Then, I built it with:
```
julia --color=yes build_tarballs.jl --verbose --deploy=staticfloat/HelloWorldC_jll.jl
```

Which resulted in wrappers that have the relevant piece of init code: https://github.com/staticfloat/HelloWorldC_jll.jl/blob/f7788fa180b68067edc305755df8d20617bb0a9f/src/wrappers/x86_64-apple-darwin14.jl#L58

Looks good to me, merging!

**Pull request merged**